Android/tun2socks/build.gradle

```diff
@@ -1,2 +1,77 @@
-configurations.maybeCreate("default")
-artifacts.add("default", file('tun2socks.aar'))
\ No newline at end of file
+// We use Android library plugin to get the Android SDK path.
+plugins {
+    id('com.android.library')
+}
+
+android {
+    compileSdkVersion 33
+    defaultConfig {
+        minSdkVersion 16
+    }
+}
+
+configurations {
+    aarBinary {
+        canBeConsumed = true
+        canBeResolved = false
+    }
+}
+
+def goBuildDir = file("${buildDir}/go")
+def outputAAR = file("${buildDir}/tun2socks.aar")
+
+def srcDir = "${rootDir}/tun2socks/intra"
+def srcPackages = [srcDir,
+                   "${srcDir}/android",
+                   "${srcDir}/doh",
+                   "${srcDir}/split",
+                   "${srcDir}/protect"]
+
+// Make sure that the go build directory exists.
+task ensureBuildDir() {
+    doLast {
+        goBuildDir.mkdirs()
+    }
+}
+
+// Install `gomobile` and `gobind` to the build directory, so that the user
+// does not need to install them on their system or call `gomobile init`.
+task ensureGoMobile(type: Exec, dependsOn: ensureBuildDir) {
+    // Define outputs so this task will only be executed when they don't exist
+    outputs.file("${goBuildDir}/gomobile")
+    outputs.file("${goBuildDir}/gobind")
+
+    environment 'GOBIN', goBuildDir.getPath()
+
+    commandLine('go', 'install',
+                'golang.org/x/mobile/cmd/gomobile@latest',
```

Comment:
If you use `@latest`, you will ignore the go.mod, and you won't get consistent versioning.

Use `go build` instead, so it's aware of the module: https://github.com/Jigsaw-Code/outline-sdk/tree/main/x/mobileproxy#build-the-go-mobile-binaries-with-go-build

Reply:
Removing `@latest` from the package name would do the work as well. If the `@` symbol is omitted, `go install` will also read the `go.mod` file.

Here is the source code that handles package names containing `@` in `go install`: https://github.com/golang/go/blob/ad9e6edfddf7c68b7a549ab7b491919f0980889d/src/cmd/go/internal/work/build.go#L690-L693

Reply:
The code is not very clear if it will install in module mode and use the pinned dependencies. 
But the documentation seems to imply it will be installed in module-aware mode. So I guess it's fine to just remove the version suffix.

I still think go build is superior, because you don't need to know the rules of where Go installs things. You can explicitly specify the output location with the ` -o`  flag. This is especially helpful for people not used to Go, since otherwise what happens to the binary is just magic. But it's up to you.

Reply:
replaced it with `go build` since `go install` won't install reproducible binaries, either.